### DISCUSSION TITLE
DPC++ capture this pointer in the kernal function

### DISCUSSION THREAD

**Discussion by Maccchiatooo** [uxlfoundation/oneDPL#585] [General]
When I want to capture the this pointer or outside variable in DPC++ code, I have to write it in my lambda function explicitly

```
    class abc{
    queue Q{};
    std::array<int, constsize> e;
    std::array<double, constsize>t;
    abc(){
    ua = malloc_shared<double>(local_size, this->Q);
    }
    
    void b();
    
  }

 void abc::b(){
      for(int i=0;i<constsize;i++){
      e[i]=i;
      t[i]=2*i; 
      }

    buffer<int> ee{e};
    buffer<double> tt{t};
auto ini2 = this->Q.submit([&,uq=this->ua](handler &h)
            { accessor eee{ee, h, read_only};
              accessor ttt{tt, h, read_only};
                                        
              h.parallel_for(range{size1, size2, size3}, [=](id<3> idx)
              double eu=uq[id[0]];
              int aa=eee[id[1]];
              double cc=ttt[id[2]];
             }

 }
```
But, now I have a lot of variables that need to be captured in the lambda function, do I have to use the same method to write them explicitly? Or there are some other tricks to make this?

**Comment by reble** [comment]
Is a by-copy capture with exceptions by-reference what you're looking for:
`[=, &size1,&size2,&size3](handler&h){};`
Btw, where are those variables `sizeX` coming from? Can't find them in the code above.
Also your code won't compile. I've created a cleaned up version here:
https://godbolt.org/z/6eenrvjzj
Let us know if you have any more questions. Happy to help.

**Comment by reble** [comment]
Converted to discussion, since it's not really a oneDPL issue.
